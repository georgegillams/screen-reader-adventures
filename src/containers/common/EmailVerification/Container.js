import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import PageTitle from 'components/common/PageTitle';
import { DebugObject } from 'components/common/DebugObject';
import { Paragraph } from 'gg-components/Paragraph';
import ErrorDisplay from 'components/common/ErrorDisplay';

import { withRouter } from 'next/router';

const EmailVerification = props => {
  const [verificationAttempted, setVerificationAttempted] = useState(false);

  const {
    router,
    verify,

    verificationState,
    authenticatorState,

    className,
  } = props;

  const { verifyError, verifyResult } = verificationState;

  useEffect(() => {
    let token = null;
    if (router && router.query) {
      token = router.query.token;
    }
    if (!verificationAttempted && token) {
      verify(token);
      setVerificationAttempted(true);
    }
  }, [verificationState, router]);

  const success = verifyResult && !verifyError;

  const outerClassNames = [];

  if (className) {
    outerClassNames.push(className);
  }

  return (
    <div className={outerClassNames.join(' ')}>
      <PageTitle name={success ? 'Email confirmed' : 'Confirming email'}>
        {success && (
          <>
            <Paragraph>Thank you for confirming your email address</Paragraph>
          </>
        )}
        <ErrorDisplay message="Something went wrong whilst confirming your email" error={verifyError}></ErrorDisplay>
      </PageTitle>
      <DebugObject
        debugTitle="EmailVerification"
        debugObject={{
          verify,
          authenticatorState,
          verificationState,
        }}
      />
    </div>
  );
};

EmailVerification.propTypes = {
  router: PropTypes.shape({
    push: PropTypes.func,
    query: PropTypes.shape({
      token: PropTypes.string,
      redirect: PropTypes.string,
    }).isRequired,
  }),
  verify: PropTypes.func.isRequired,
  verificationState: PropTypes.shape({
    token: PropTypes.string,
    verifying: PropTypes.bool,
    verifyError: PropTypes.object,
    verifyResult: PropTypes.object,
  }).isRequired,
  authenticatorState: PropTypes.shape({
    user: PropTypes.object,
  }).isRequired,
  className: PropTypes.string,
};

EmailVerification.defaultProps = {
  router: null,
  verificationState: null,
  magicLoginState: null,
  className: null,
};

export default withRouter(EmailVerification);
