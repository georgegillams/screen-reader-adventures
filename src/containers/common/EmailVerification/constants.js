export const KEY = 'email-verification';
export const VERIFY = `app/${KEY}/VERIFY`;
