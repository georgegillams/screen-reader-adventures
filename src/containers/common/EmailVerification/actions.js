import { createRoutine } from 'redux-saga-routines';

import { VERIFY } from './constants';

export const verify = createRoutine(VERIFY);
