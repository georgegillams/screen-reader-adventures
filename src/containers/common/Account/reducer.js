import produce from 'immer';

import { logout, requestVerificationEmail } from './actions';

export const initialState = {
  loggingOut: false,
  logOutError: null,
  logOutResult: null,
  requestingVerificationEmail: false,
  requestVerificationEmailError: null,
  requestVerificationEmailResult: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case logout.REQUEST:
        draft.loggingOut = true;
        draft.logOutError = null;
        break;

      case logout.SUCCESS:
        draft.loggingOut = false;
        draft.logOutResult = payload;
        break;

      case logout.FAILURE:
        draft.loggingOut = false;
        draft.logOutError = payload;
        break;

      case requestVerificationEmail.REQUEST:
        draft.requestingVerificationEmail = true;
        draft.requestVerificationEmailError = null;
        break;

      case requestVerificationEmail.SUCCESS:
        draft.requestingVerificationEmail = false;
        draft.requestVerificationEmailResult = payload;
        break;

      case requestVerificationEmail.FAILURE:
        draft.requestingVerificationEmail = false;
        draft.requestVerificationEmailError = payload;
        break;
    }
  });

export default reducer;
