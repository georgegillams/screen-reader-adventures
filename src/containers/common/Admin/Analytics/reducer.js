import produce from 'immer';

import { load } from './actions';

export const initialState = {
  analytics: null,
  loading: false,
  loadError: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case load.REQUEST:
        draft.loading = true;
        draft.loadError = null;
        break;

      case load.SUCCESS:
        draft.loading = false;
        draft.analytics = payload;
        break;

      case load.FAILURE:
        draft.loading = false;
        draft.loadError = payload;
        break;
    }
  });

export default reducer;
