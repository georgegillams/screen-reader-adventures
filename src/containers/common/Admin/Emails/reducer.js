import produce from 'immer';

import { load, resend } from './actions';

export const initialState = {
  emails: null,
  loading: false,
  loadError: null,

  emailToResend: null,
  resending: false,
  resendError: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case load.REQUEST:
        draft.loading = true;
        draft.loadError = null;
        break;

      case load.SUCCESS:
        draft.loading = false;
        draft.emails = payload;
        break;

      case load.FAILURE:
        draft.loading = false;
        draft.loadError = payload;
        break;

      case resend.TRIGGER:
        draft.emailToResend = payload.emailToResend;
        break;

      case resend.REQUEST:
        draft.resending = true;
        draft.resendError = null;
        break;

      case resend.SUCCESS:
        draft.resending = false;
        break;

      case resend.FAILURE:
        draft.resending = false;
        draft.resendError = payload;
        break;
    }
  });

export default reducer;
