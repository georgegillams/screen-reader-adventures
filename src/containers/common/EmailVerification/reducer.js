import produce from 'immer';

import { verify } from './actions';

export const initialState = {
  token: null,
  verifying: false,
  verifyError: null,
  verifyResult: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case verify.TRIGGER:
        draft.token = payload;
        break;

      case verify.REQUEST:
        draft.verifying = true;
        draft.verifyError = null;
        break;

      case verify.SUCCESS:
        draft.verifying = false;
        draft.verifyResult = payload;
        break;

      case verify.FAILURE:
        draft.verifying = false;
        draft.verifyError = payload;
        break;
    }
  });

export default reducer;
