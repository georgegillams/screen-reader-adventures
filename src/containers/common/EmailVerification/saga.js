import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { selectState } from './selectors';
import { verify } from './actions';
import { loadAuth } from 'containers/common/Authenticator/actions';

export function* doVerify() {
  const currentState = yield select(selectState());
  const { token } = currentState;
  const requestURL = apiStructure.verifyEmail.fullPath;

  try {
    yield put(verify.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify({ verificationKey: token }),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(verify.failure(result));
    } else {
      yield put(verify.success(result));
      yield put(loadAuth.trigger());
    }
  } catch (err) {
    yield put(verify.failure(err));
  } finally {
    yield put(verify.fulfill());
  }
}
export default function* saga() {
  yield takeLatest(verify.TRIGGER, doVerify);
}
