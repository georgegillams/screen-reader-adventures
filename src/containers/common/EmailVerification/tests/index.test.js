import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { initialState } from '../reducer';
import { initialState as initialAuthenticatorState } from 'containers/common/Authenticator/reducer';

import configureStore from 'client-utils/common/redux/configure-store';

import EmailVerificationIndex from '../index';
import EmailVerification from '../Container';

describe('<EmailVerification />', () => {
  let store;
  const spy = jest.fn();

  beforeAll(() => {
    store = configureStore({});
  });

  it('should render correctly - index', () => {
    const { container } = render(
      <Provider store={store}>
        <EmailVerificationIndex />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with initial state', () => {
    const { container } = render(
      <Provider store={store}>
        <EmailVerification
          verify={spy}
          verificationState={{
            ...initialState,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with verifying=true', () => {
    const { container } = render(
      <Provider store={store}>
        <EmailVerification
          verify={spy}
          verificationState={{
            ...initialState,
            verifying: true,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with verifyError', () => {
    const { container } = render(
      <Provider store={store}>
        <EmailVerification
          verify={spy}
          verificationState={{
            ...initialState,
            verifyError: { error: 'not_found', errorMessage: 'Invalid verification link' },
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with user', () => {
    const { container } = render(
      <Provider store={store}>
        <EmailVerification
          verify={spy}
          verificationState={{
            ...initialState,
            verifyResult: { success: 'Email verified' },
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
            user: { name: 'userName' },
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
