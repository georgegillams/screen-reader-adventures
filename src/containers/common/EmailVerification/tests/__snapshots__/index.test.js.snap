// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmailVerification /> should render correctly - index 1`] = `
<div>
  <div
    class=""
  >
    <h1
      class="page-title__heading"
    >
      Confirming email
    </h1>
  </div>
</div>
`;

exports[`<EmailVerification /> should render correctly with initial state 1`] = `
<div>
  <div
    class=""
  >
    <h1
      class="page-title__heading"
    >
      Confirming email
    </h1>
  </div>
</div>
`;

exports[`<EmailVerification /> should render correctly with user 1`] = `
<div>
  <div
    class=""
  >
    <h1
      class="page-title__heading"
    >
      Email confirmed
    </h1>
    <span
      class="paragraph"
    >
      Thank you for confirming your email address
    </span>
  </div>
</div>
`;

exports[`<EmailVerification /> should render correctly with verifyError 1`] = `
<div>
  <div
    class=""
  >
    <h1
      class="page-title__heading"
    >
      Confirming email
    </h1>
    <span
      class="paragraph"
    >
      <span
        class="error"
      >
        Something went wrong whilst confirming your email
        <br />
        Invalid verification link
      </span>
    </span>
  </div>
</div>
`;

exports[`<EmailVerification /> should render correctly with verifying=true 1`] = `
<div>
  <div
    class=""
  >
    <h1
      class="page-title__heading"
    >
      Confirming email
    </h1>
  </div>
</div>
`;
